app/assets/javascripts/discourse/app/lib/plugin-api.js

```diff
@@ -1882,6 +1883,49 @@ class PluginApi {
   registerNotificationTypeRenderer(notificationType, func) {
     registerNotificationTypeRenderer(notificationType, func);
   }
+
+  /**
+   * EXPERIMENTAL. Do not use.
+   * Registers a new tab in the user menu. This API method expects a callback
+   * that should return a class inheriting from the class (UserMenuTab) that's
+   * passed to the callback. See discourse/app/lib/user-menu/tab.js for
+   * documentation of UserMenuTab.
+   *
+   * ```
+   * api.registerUserMenuTab((UserMenuTab) => {
+   *   return class extends UserMenuTab {
+   *     get id() {
+   *       return "custom-tab-id";
+   *     }
+   *
+   *     get shouldDisplay() {
+   *       return this.siteSettings.enable_custom_tab && this.currentUser.admin;
+   *     }
+   *
+   *     get count() {
+   *       return this.currentUser.my_custom_notification_count;
+   *     }
+   *
+   *     get panelComponent() {
+   *       return "your-custom-glimmer-component";
```

Comment:
Is there a way we can restrict what is customizable in each panel? I think it'll be better if we start with some constraints first instead of letting people display anything they want.

Reply:
Also, I think it'll be helpful to state what is being passed to the component when it is rendered because I think it is very unclear what the component is rendered with at the moment. Or is the expectation for each component to fetch its own data?

Reply:
OK just had a closer look at the API. I think we shouldn't just accept any component but rather only accept components which inherits from `UserMenuItemsList`.

Reply:
Hmm what does restricting it to subclasses of `UserMenuItemsList` achieve? A child component can render its own template and isn't required to use its parent's template, and also as far as I can tell we can't lookup a component's class from its name so I don't think can check the ancestors of the component we're given.

Reply:
I'm relooking at the API at the moment because I think we're being too flexibile here. The contents of each tab is essentially a list so my ideal implementation is one component for the tab, one component for the panel and one component for each item in the panel.

Reply:
Just to give some visual context, the `panelComponent` from this API is rendered in the red square:

<img src="https://user-images.githubusercontent.com/17474474/183368055-f5897130-7258-4eeb-a20c-374f58be9131.png" width=400>

Reply:
Yup I understand but I'm not sure if we need to give plugins all the flexibility here. Right now, we're basically giving the plugins a blank box to insert whatever they want and I'm not sure that is ideal.

Reply:
I don't think a sweet spot exists between 'too flexible' and 'too restrictive'. If we never want to allow plugins to insert random stuff, then we can't accept Glimmer components from them for anything and we have to stick to regular classes that provide pieces of data that we render, but in my opinion that's too restrictive. I think it's totally reasonable for plugins to want to add, say, 2 or 3 buttons/links at the bottom of the panel or maybe have a header at the top. I worry about introducing a very restrictive API that we then have to loosen up for an official plugin.

Taking a step back here, for what reasons do you want to avoid flexibility here? Is it out of concern that it'd be hard to migrate to another rendering system/engine?

Reply:
>  for what reasons do you want to avoid flexibility here? Is it out of concern that it'd be hard to migrate to another rendering system/engine?

Because we don't need the flexibility yet and it makes the API really hard to use. 

Regarding flexibility, I don't really see a good reason for us to start from a point where we give plugin authors all the flexibility when we don't even know the possible use cases yet. Instead, I think makes our API more maintainable if we can start with something restrictive based on the actual use cases we need right now and then slowly expand the API over time. 

The second is with regards to the ease of usage for the API. I tried it while working on the chat plugin and it feels too complex. We have to register a `UserMenuTab` first and then implement a component which needs to act like a certain component before registering the notifications type. It feels like too many separate pieces right now. Instead, I think adding a tab and the panel should be a single API interface.

Reply:
Anyway, I don't know how the API should look like at the moment but I'm experimenting with something locally and will tag you once I'm done.

Reply:
Ok then, I'll convert this PR to draft for the time being. 

> experimenting with something locally

Keep in mind that whatever API we come up with needs to work with the assign's plugin use-case which has to render a list of assignments in the panel.

Reply:
@OsamaSayegh Just to confirm though, as of right now everything in the panel is just a `<ul>` with a `<li>` right?